src/transport/raw/TCP.cpp

```diff
@@ -162,75 +165,58 @@ Inet::TCPEndPoint * TCPBase::FindActiveConnection(const PeerAddress & address)
-    const size_t prefixSize      = header.EncodeSizeBytes() + kPacketSizeBytes;
-    Inet::TCPEndPoint * endPoint = nullptr;
-    uint16_t actualEncodedHeaderSize;
+    const size_t prefixSize = header.EncodeSizeBytes() + kPacketSizeBytes;
 
-    VerifyOrExit(address.GetTransportType() == Type::kTcp, err = CHIP_ERROR_INVALID_ARGUMENT);
-    VerifyOrExit(mState == State::kInitialized, err = CHIP_ERROR_INCORRECT_STATE);
-    VerifyOrExit(prefixSize + msgBuf->DataLength() <= std::numeric_limits<uint16_t>::max(), err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(address.GetTransportType() == Type::kTcp, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(mState == State::kInitialized, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(prefixSize + msgBuf->DataLength() <= std::numeric_limits<uint16_t>::max(), CHIP_ERROR_INVALID_ARGUMENT);
 
     // The check above about prefixSize + msgBuf->DataLength() means prefixSize
     // definitely fits in uint16_t.
-    VerifyOrExit(msgBuf->EnsureReservedSize(static_cast<uint16_t>(prefixSize)), err = CHIP_ERROR_NO_MEMORY);
-
-    {
-        msgBuf->SetStart(msgBuf->Start() - prefixSize);
+    VerifyOrReturnError(msgBuf->EnsureReservedSize(static_cast<uint16_t>(prefixSize)), CHIP_ERROR_NO_MEMORY);
 
-        uint8_t * output = msgBuf->Start();
+    msgBuf->SetStart(msgBuf->Start() - prefixSize);
 
-        // Length is actual data, without considering the length bytes themselves
-        VerifyOrExit(msgBuf->DataLength() >= kPacketSizeBytes, err = CHIP_ERROR_INTERNAL);
+    // Length is actual data, without considering the length bytes themselves
+    VerifyOrReturnError(msgBuf->DataLength() >= kPacketSizeBytes, CHIP_ERROR_INTERNAL);
 
-        LittleEndian::Write16(output, static_cast<uint16_t>(msgBuf->DataLength() - kPacketSizeBytes));
+    uint8_t * output = msgBuf->Start();
+    LittleEndian::Write16(output, static_cast<uint16_t>(msgBuf->DataLength() - kPacketSizeBytes));
 
-        err = header.Encode(output, msgBuf->DataLength(), &actualEncodedHeaderSize, payloadFlags);
-        SuccessOrExit(err);
+    uint16_t actualEncodedHeaderSize;
+    ReturnErrorOnFailure(header.Encode(output, msgBuf->DataLength(), &actualEncodedHeaderSize, payloadFlags));
 
-        // header encoding has to match space that we allocated
-        VerifyOrExit(prefixSize == actualEncodedHeaderSize + kPacketSizeBytes, err = CHIP_ERROR_INTERNAL);
-    }
+    // header encoding has to match space that we allocated
+    VerifyOrReturnError(prefixSize == actualEncodedHeaderSize + kPacketSizeBytes, CHIP_ERROR_INTERNAL);
 
     // Reuse existing connection if one exists, otherwise a new one
     // will be established
-
-    endPoint = FindActiveConnection(address);
+    Inet::TCPEndPoint * endPoint = FindActiveConnection(address);
 
     if (endPoint != nullptr)
     {
-        err = endPoint->Send(msgBuf);
+        return endPoint->Send(autofree.Release());
     }
     else
```

Comment:
What's our general position on else after return?

Reply:
I generally prefer single point of return or exit / error handling (RAII or not) which moots this particular facet of coding style discussion.

Reply:
I have created this PR after seeing that mandating goto with single point of exit can result in significantly hard to follow code. 

The premise is that we will have the option to use return if the code is cleaner as a result. Mandating exit label always seems too strong of a requirement.

Reply:
Regarding the else after return: I think generally I would say no else is needed after return. In this particular case however, this is a final statement that reads like:

```cpp
return (endpoint != nullptr)
    ? endPoint->Send(autofree.Release())
    : SendAfterConnect(address, autofree.Release());
```

so I thought else helps clarity. I would drop the else if the readability reviews would ask for it.

Reply:
+1 that mixing in a single method would be absolutely bad - you get the benefit of neither. I am not arguing for that. I am arguing that I should be allowed to use return if it makes the code easier/clearer to read.

Reply:
I could not find a good way to count, but `find . -type f | egrep -v '/third_party/' | xargs egrep -zc 'exit:\s*return' | grep -v ':0' | wc -l` seems to indicate that there are 149 instances today where 'exit: return...' is used. I believe those instances would benefit from just using return in the first place.

Reply:
C forces you into the goto exit pattern, but C++ doesn't, and the goto style with declarations front loaded seems unusual for C++. Google's style guide doesn't even bother to recommend against using goto because it is so uncommon in the first place [1].

[1] https://google.github.io/styleguide/cppguide.html#Goals